second_pass.c

```diff
@@ -142,258 +142,237 @@ void read_line2(char *line, FILE *ob_file, FILE *ent_file, FILE *ext_file, int *
+			immed = check_addressing(to_comma(line), error);
+			addressing2 = check_addressing(to_comma(to_comma(line,error),error));
+			to_inst(23,addressing1,addressing2,IRELEVANT,immed,IRELEVANT,IRELEVANT,ic);
+			(*ic++);
+			break;
+
+			case 'sh':
+			addressing1 = check_addressing(line, error);
+			immed = check_addressing(to_comma(line), error);
+			addressing2 = check_addressing(to_comma(to_comma(line,error),error));
+			to_inst(24,addressing1,addressing2,IRELEVANT,immed,IRELEVANT,IRELEVANT,ic);
+			(*ic++);
+			break;
+
+			case 'jmp':
+			addressing1 = check_addressing(next_word(line), error);
+			if(is_label(next_word(line)))
+				if(is_dir(next_word(line)) == "entry")
+					to_inst(30,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,addressing1,ic);
+				else
+					to_inst(30,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,ic);
 			else
-			{
-				to_inst(LABEL_TYPE, instindex
-	, addressing1, addressing2, symbol->address, RELOCATABLE, ic);
-				(*ic)++;
-			}
-		}
-		
-		/* if second operand is index */
+				to_inst(30,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,REG_ADDRESS,addressing1,ic);
+			(*ic++);
+			break;
 
-		if(addressing2 == INDEX_ADDRESS)
-		{
-			/*check if macro */
-			char key[MAX_MACRO_LEN];
-			int j=1;
-			symbol_node * node;
-			for(i=0;*line!='[';i++)
-			{
-				label[i] = *line;
-				line++;
-			}
-			label[i] = '\0';
-			symbol = search_sym(label);
-			while(line[j] != ']' && line[j] != '\0')
-			{
-				key[j-1] = line[j];
-				j++;
-			}
-			key[j-1] = '\0';
-			if((node = search_sym(key)) != NULL)
-			{
-				*(line+1) = node->address + '0';
-			}
+			case 'la':
+			addressing1 = check_addressing(next_word(line), error);
+			if(is_dir(next_word(line)) == "entry")
+				to_inst(31,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,addressing1,ic);
+			else
+				to_inst(31,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,REG_ADDRESS,0,ic);
+			(*ic++);
+			break;
 
-			if(symbol == NULL)
-			{
-				*error = NDEF_LABEL_ERROR;
-				return;
-			}
+			case 'call':
+			addressing1 = check_addressing(next_word(line), error);
+			immed = check_addressing(to_comma(line), error);
+			if(is_dir(next_word(line)) == "entry")
+				to_inst(32,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,addressing1,ic);
+			else
+				to_inst(32,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,ic);
+			(*ic++);
+			break;
 
+			case 'stop':
+			to_inst(63,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT,IRELEVANT);
+			(*ic++);
+			break;
 
-		/* if the label is external */
-		else if(symbol->ext_flag == TRUE)
-		{
-			to_inst(LABEL_TYPE, instindex
-, addressing1, addressing2, FALSE, EXTERNAL, ic);
-			(*ic)++;
 		}
-		else
-		{
-			to_inst(LABEL_TYPE, instindex
-, addressing1, addressing2, symbol->address, RELOCATABLE, ic);
-			(*ic)++;
-		}
-		line ++;
-		data = *line - '0';
-
-		to_inst(DATA_TYPE, instindex, reg_src, reg_dst, data, ABSOLUTE, ic);
-		(*ic)++;
+			line = to_comma(line);
```

Comment:
why you need to comma? 
shouldn't you go to next line? or next word? and check if it's the last word?